storage/client/src/main/java/com/azure/storage/blob/BlockBlobAsyncClient.java

```diff
@@ -134,33 +150,63 @@ public static BlockBlobClientBuilder builder() {
      *         {@link Metadata}
      * @param accessConditions
      *         {@link BlobAccessConditions}
-     * @param context
-     *         {@code Context} offers a means of passing arbitrary data (key/value pairs) to an
-     *         {@link com.azure.core.http.HttpPipeline}'s policy objects. Most applications do not need to pass
-     *         arbitrary data to the pipeline and can pass {@code Context.NONE} or {@code null}. Each context object is
-     *         immutable. The {@code withContext} with data method creates a new {@code Context} object that refers to
-     *         its parent, forming a linked list.
      *
      * @return
      *      A reactive response containing the information of the uploaded block blob.
      */
-    public Mono<Void> upload(Flux<ByteBuffer> data, long length, BlobHTTPHeaders headers,
-            Metadata metadata, BlobAccessConditions accessConditions, Context context) {
+    public Mono<Response<BlockBlobItem>> upload(Flux<ByteBuffer> data, long length, BlobHTTPHeaders headers,
+            Metadata metadata, BlobAccessConditions accessConditions) {
         return blockBlobAsyncRawClient
-            .upload(data.map(Unpooled::wrappedBuffer), length, headers, metadata, accessConditions, context)
-            .then();
+            .upload(data.map(Unpooled::wrappedBuffer), length, headers, metadata, accessConditions)
+            .map(rb -> new SimpleResponse<>(rb, new BlockBlobItem(rb.deserializedHeaders())));
     }
 
     public Mono<Void> uploadFromFile(String filePath) {
-        return this.uploadFromFile(filePath, null, null, null, null);
+        return this.uploadFromFile(filePath, null, null, null);
     }
 
     public Mono<Void> uploadFromFile(String filePath, BlobHTTPHeaders headers, Metadata metadata,
-            BlobAccessConditions accessConditions, Context context) {
-        //TODO make this method smart
-        return this.blockBlobAsyncRawClient
-            .upload(ByteBufFlux.fromPath(Paths.get(filePath)), new File(filePath).length())
-            .then();
+                                     BlobAccessConditions accessConditions) {
+        AsynchronousFileChannel channel;
+        try {
+            channel = AsynchronousFileChannel.open(Paths.get(filePath), StandardOpenOption.READ);
+        } catch (IOException e) {
+            return Mono.error(e);
+        }
+        final SortedMap<Long, String> blockIds = new TreeMap<>();
+        return Flux.fromIterable(sliceFile(filePath))
+            .doOnNext(chunk -> blockIds.put(chunk.offset(), getBlockID()))
+            .flatMap(chunk -> {
```

Comment:
I know you mentioned that concatMapEager in Rx had some weird performance behavior. Is the same true of flatMapSequential in Reactor? I think it might be preferable to not have state external to the reactive pipeline, but that's kind of moot.

Reply:
Theoretically it will work - but I was worried about its buffering so decided to go the dumb but works-for-sure way. We can test it more and see if flatMapSequential helps simply this after preview 1.